examples/island/Island.cpp

```diff
@@ -0,0 +1,579 @@
+
+    // Thread pool parameters
+    const unsigned int threadCount = 4;
+    const unsigned int blockCount = 32;
+
+    struct WorkItem
+    {
+        sf::Vertex* targetBuffer;
+        unsigned int index;
+    };
+
+    std::deque<WorkItem> workQueue;
+    std::vector<sf::Thread*> threads;
+    int pendingWorkCount = 0;
+    bool workPending = true;
+    bool bufferUploadPending = false;
+    sf::Mutex workQueueMutex;
+
+    struct Setting
+    {
+        const char* name;
+        float* value;
+    };
+
+    // Terrain noise parameters
+    const int perlinOctaves = 3;
+
+    float perlinFrequency = 7.0f;
+    float perlinFrequencyBase = 4.0f;
+
+    // Terrain generation parameters
+    float heightBase = 0.0f;
+    float edgeFactor = 0.9f;
+    float edgeDropoffExponent = 1.5f;
+
+    float snowcapHeight = 0.6f;
+
+    // Terrain lighting parameters
+    float heightFactor = windowHeight / 2.0f;
+    float heightFlatten = 3.0f;
+    float lightFactor = 0.7f;
+}
+
+
+// Forward declarations of the functions we define further down
+void threadFunction();
+void generateTerrain(sf::Vertex* vertexBuffer);
+
+
+////////////////////////////////////////////////////////////
+/// Entry point of application
+///
+/// \return Application exit code
+///
+////////////////////////////////////////////////////////////
+int main()
+{
+    // Create the window of the application
+    sf::RenderWindow window(sf::VideoMode(windowWidth, windowHeight), "SFML Island",
+                            sf::Style::Titlebar | sf::Style::Close);
+    window.setVerticalSyncEnabled(true);
+
+    sf::Font font;
+    if (!font.loadFromFile("resources/sansation.ttf"))
+        return EXIT_FAILURE;
+
+    bool prerequisitesSupported = sf::VertexBuffer::isAvailable() && sf::Shader::isAvailable();
+
+    // Create all of our graphics resources
+    sf::Text hudText("", font);
+    sf::Text statusText(prerequisitesSupported ? "Generating Terrain..." : "Shaders and/or Vertex Buffers Unsupported", font);
+    sf::VertexBuffer terrain(sf::Triangles, sf::VertexBuffer::Static);
+    sf::Shader terrainShader;
+    sf::Texture terrainTexture;
+    sf::RenderStates terrainStates;
+
+    // Staging buffer for our terrain data that we will upload to our VertexBuffer
+    std::vector<sf::Vertex> terrainStagingBuffer(resolutionX * resolutionY * 6);
+
+    // Create our VertexBuffer with enough space to hold the terrain geometry
+    terrain.create(resolutionX * resolutionY * 6);
+
+    // Set up our graphics resources
+    if (prerequisitesSupported)
+    {
+        statusText.setOutlineColor(sf::Color::Black);
+        statusText.setOutlineThickness(2.0f);
+
+        hudText.setCharacterSize(14);
+        hudText.setFillColor(sf::Color::White);
+        hudText.setOutlineColor(sf::Color::Black);
+        hudText.setOutlineThickness(2.0f);
+        hudText.setPosition(5.0f, 5.0f);
+
+        if (!terrainShader.loadFromFile("resources/terrain.vert", "resources/terrain.frag"))
```

Comment:
I'd move this to line 92 so that when calling `generateTerrain` the prerequisites are always met.